### DISCUSSION TITLE
convert mesh to pointcloud

### DISCUSSION THREAD

**Discussion by monajalal** [pyvista/pyvista#4613] [Q&A]
### Describe the feature you would like to be added.

I have a mesh with its texture file. Is there a method to convert this mesh to npy point cloud file such that the pcd only contains samples from the surface of the mesh and we could have control over the number of samples? Let's say I only want to have a point cloud npy with 1000 points.

Also, will I be able to capture the model color info besides its xyz?
These are the files I have:
```
-rw-rw-r-- 1 mona mona  224 Apr 25 15:12 textured_simple_1000.obj.mtl
-rw-rw-r-- 1 mona mona 1.6M Apr 25 15:12 textured_simple_1000.obj
-rwxrwxr-x 1 mona mona 8.1M Apr 25 15:12 texture_map.png
```
Also, I notice when I open the obj file, it doesn't render the mesh with it while with MeshLab I am able to see it rendered.

Here's what I see:
![Screenshot from 2023-06-07 13-44-48](https://github.com/pyvista/pyvista/assets/1892917/536d2232-3a4c-4e20-b738-f6807a7c6992)

```
import pyvista as pv

# Load the 3D model from an OBJ file
mesh = pv.read('/home/mona/zephyr/notebooks/meshes/spam/spam/textured_simple_1000.obj')

mesh.plot()
```
```
points = pv.PointSet(mesh.points)
```

gives me:
![Screenshot from 2023-06-07 13-58-00](https://github.com/pyvista/pyvista/assets/1892917/87d779f6-2be6-4851-8347-f5ba5a6782ab)

![Screenshot from 2023-06-07 14-03-15](https://github.com/pyvista/pyvista/assets/1892917/c2e147fa-890b-4890-a837-4ae520efd882)

but I want to sample only 1000 points using pyvista I guess.

### Pseudocode or Screenshots

```python
Insert code here
```
Or drag your screenshot here!

**Comment by MatthewFlamm** [comment]
Are you looking to sample from the nodal points of the mesh or anywhere on the surface?

**Comment by MatthewFlamm** [comment]
I started pyransame https://matthewflamm.github.io/pyransame/ to solve the latter problem, i.e.  sampling of the area of the mesh.  If you are looking to accomplish the former, i.e. sampling the nodal points of the mesh, there is also an example in the docs on how to use `np.random` to accomplish this.

**Comment by monajalal** [comment]
so I got this
```
points = pv.PointSet(mesh.points)
pointsets_np = np.array(points)
```
However, the pointsets_np is rather weird not a normal ndarray

![Screenshot from 2023-06-07 16-38-49](https://github.com/pyvista/pyvista/assets/1892917/fc9413a3-c9e1-40f0-b7f4-7aa0ea7f36bd)

and anyways I am not able to use the method I use for it
```
from mpl_toolkits.mplot3d import Axes3D



# Create a 3D plot
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

# Plot the point cloud data
ax.scatter(model_data['model_points'][:,0], model_data['model_points'][:,1], model_data['model_points'][:,2], s=1)

# Set the axis labels
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')

# Show the plot
plt.show()
```

**Comment by MatthewFlamm** [comment]
`mesh.points` is already a numpy ndarray:

```
import pyvista as pv
import numpy as np
mesh = pv.Sphere()
isinstance(mesh.points, np.ndarray) # True
```

**Comment by monajalal** [comment]
@MatthewFlamm I also got this one

```
import pyvista as pv
import numpy as np

# Load the 3D model from an OBJ file
mesh = pv.read('/home/mona/zephyr/notebooks/meshes/spam/spam/textured_simple_1000.obj')

# Create a point cloud from the mesh points
points = pv.PolyData(mesh.points)

# Generate a scalar array
scalar = np.random.rand(points.n_points)

# Visualize the point cloud with scalar coloring
points.plot(scalars=scalar)
```

![Screenshot from 2023-06-07 16-47-35](https://github.com/pyvista/pyvista/assets/1892917/b85be871-40d1-4bd0-bb4b-6b72ab417ce2)


What do you think about this?

I was not sure if the points were sampled only from the surface?
Also, is there a way for the point cloud points to capture the color of texture map?


![texture_map](https://github.com/pyvista/pyvista/assets/1892917/8bde3ff8-b062-4271-879e-0a157deed468)


I have attached the texture map of the SPAM here for you but it also has an mtl file

**Comment by monajalal** [comment]
as you see my pointcloud (the green one) has so many points I was hoping to reduce number of points

**Comment by MatthewFlamm** [comment]
What is the purpose of this operation?  It seems you are trying to plot the mesh with a texture, which all can be done within PyVista. But you are asking about plotting this in matplotlib.

However, I'm not that familiar with the texture operations in PyVista so hopefully others can chime in here.

**Comment by MatthewFlamm** [comment]
I suspect you really need an answer to a different question.  But if you need to sample the number of points on the mesh, one option is to use the methodology in the link provided above to do so.  For example (mostly from [pyransame docs](https://matthewflamm.github.io/pyransame/)).  You could modify it to grab the point ids instead of the point locations too.

```py
import numpy as np
import pyvista as pv
from pyvista import examples
mesh = examples.download_antarctica_velocity()
vertices = np.random.default_rng().choice(mesh.points, 500)  # Only choose 500 nodal points randomly
```

This will give you a uniform sampling, but it will be uneven in space by the nature of random sampling.  If you are looking for this aspect, i.e. a sampling of points that is relatively equally spaced, there is a vtk Filter `vtkMaskPoints` that is not yet implemented in PyVista to do this.  This would be a feature request for PyVista or you could implement it through `vtk` in the meantime.
